Mistral model requires flash attention v2

Hi 
I got the following error:
```
NotImplementedError: Mistral model requires flash attention v2
```

I tried use model=TheBloke/Mistral-7B-Instruct-v0.1-GPTQ

Thanks

Hi @tgaddair , here is the full logs

```
model=TheBloke/Mistral-7B-Instruct-v0.1-GPTQ

docker run --gpus all --shm-size 1g -p 8080:80 -v $volume:/data -e HF_HUB_ENABLE_HF_TRANSFER=0 ghcr.io/predibase/lorax:latest --model-id $model



2023-11-17T02:38:15.469298Z  INFO lorax_launcher: Args { model_id: "TheBloke/Mistral-7B-Instruct-v0.1-GPTQ", adapter_id: "", source: "hub", adapter_source: "hub", revision: None, validation_workers: 2, sharded: None, num_shard: None, quantize: None, dtype: None, trust_remote_code: false, max_concurrent_requests: 128, max_best_of: 2, max_stop_sequences: 4, max_input_length: 1024, max_total_tokens: 2048, waiting_served_ratio: 1.2, max_batch_prefill_tokens: 4096, max_batch_total_tokens: None, max_waiting_tokens: 20, hostname: "f5ab61aeb3c5", port: 80, shard_uds_path: "/tmp/lorax-server", master_addr: "localhost", master_port: 29500, huggingface_hub_cache: Some("/data"), weights_cache_override: None, disable_custom_kernels: false, cuda_memory_fraction: 1.0, json_output: false, otlp_endpoint: None, cors_allow_origin: [], watermark_gamma: None, watermark_delta: None, ngrok: false, ngrok_authtoken: None, ngrok_edge: None, env: false, download_only: false }
2023-11-17T02:38:15.469394Z  INFO download: lorax_launcher: Starting download process.
2023-11-17T02:38:18.625603Z  INFO lorax_launcher: hub.py:123 Download file: model.safetensors

2023-11-17T02:44:40.962134Z  INFO lorax_launcher: hub.py:131 Downloaded /data/models--TheBloke--Mistral-7B-Instruct-v0.1-GPTQ/snapshots/6ae1e4ae2cfbaf107c705ed722ec243b4f88014d/model.safetensors in 0:06:22.

2023-11-17T02:44:40.962245Z  INFO lorax_launcher: hub.py:153 Download: [1/1] -- ETA: 0

2023-11-17T02:44:41.505805Z  INFO download: lorax_launcher: Successfully downloaded weights.
2023-11-17T02:44:41.505949Z  INFO shard-manager: lorax_launcher: Starting shard rank=0
2023-11-17T02:44:43.908316Z  WARN lorax_launcher: flash_attn.py:48 Unable to use Flash Attention V2: GPU with CUDA capability 7 5 is not supported for Flash Attention V2

2023-11-17T02:44:43.911488Z  WARN lorax_launcher: __init__.py:73 Could not import Mistral model: Mistral model requires flash attn v2

2023-11-17T02:44:44.480708Z ERROR lorax_launcher: server.py:218 Error when initializing model
Traceback (most recent call last):
  File "/opt/conda/bin/lorax-server", line 8, in <module>
    sys.exit(app())
  File "/opt/conda/lib/python3.9/site-packages/typer/main.py", line 311, in __call__
    return get_command(self)(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/typer/core.py", line 778, in main
    return _main(
  File "/opt/conda/lib/python3.9/site-packages/typer/core.py", line 216, in _main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/typer/main.py", line 683, in wrapper
    return callback(**use_params)  # type: ignore
  File "/opt/conda/lib/python3.9/site-packages/lorax_server/cli.py", line 81, in serve
    server.serve(
  File "/opt/conda/lib/python3.9/site-packages/lorax_server/server.py", line 262, in serve
    asyncio.run(
  File "/opt/conda/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/conda/lib/python3.9/asyncio/base_events.py", line 634, in run_until_complete
    self.run_forever()
  File "/opt/conda/lib/python3.9/asyncio/base_events.py", line 601, in run_forever
    self._run_once()
  File "/opt/conda/lib/python3.9/asyncio/base_events.py", line 1905, in _run_once
    handle._run()
  File "/opt/conda/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
> File "/opt/conda/lib/python3.9/site-packages/lorax_server/server.py", line 214, in serve_inner
    model = get_model(
  File "/opt/conda/lib/python3.9/site-packages/lorax_server/models/__init__.py", line 283, in get_model
    raise NotImplementedError("Mistral model requires flash attention v2")
NotImplementedError: Mistral model requires flash attention v2

2023-11-17T02:44:44.910114Z ERROR shard-manager: lorax_launcher: Shard complete standard error output:

Downloading (…)lve/main/config.json: 100%|██████████| 963/963 [00:00<00:00, 46.9kB/s]
Traceback (most recent call last):

  File "/opt/conda/bin/lorax-server", line 8, in <module>
    sys.exit(app())

  File "/opt/conda/lib/python3.9/site-packages/lorax_server/cli.py", line 81, in serve
    server.serve(

  File "/opt/conda/lib/python3.9/site-packages/lorax_server/server.py", line 262, in serve
    asyncio.run(

  File "/opt/conda/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)

  File "/opt/conda/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()

  File "/opt/conda/lib/python3.9/site-packages/lorax_server/server.py", line 214, in serve_inner
    model = get_model(

  File "/opt/conda/lib/python3.9/site-packages/lorax_server/models/__init__.py", line 283, in get_model
    raise NotImplementedError("Mistral model requires flash attention v2")

NotImplementedError: Mistral model requires flash attention v2
 rank=0
Error: ShardCannotStart
2023-11-17T02:44:45.009297Z ERROR lorax_launcher: Shard 0 failed to start
2023-11-17T02:44:45.009346Z  INFO lorax_launcher: Shutting down shards

```

--- a/server/lorax_server/models/__init__.py
+++ b/server/lorax_server/models/__init__.py
@@ -43,55 +43,6 @@ __all__ = [
     "get_model",
 ]
 
-FLASH_ATT_ERROR_MESSAGE = "{} requires Flash Attention enabled models."
-
-FLASH_ATTENTION = True
-try:
-    from lorax_server.models.flash_rw import FlashRWSharded
-    from lorax_server.models.flash_neox import FlashNeoXSharded
-    from lorax_server.models.flash_llama import FlashLlama
-    from lorax_server.models.flash_gemma import FlashGemma
-    from lorax_server.models.flash_gpt2 import FlashGPT2
-    from lorax_server.models.flash_qwen import FlashQwen
-    from lorax_server.models.flash_phi import FlashPhi
-    from lorax_server.models.flash_santacoder import (
-        FlashSantacoderSharded,
-    )
-
-except ImportError as e:
-    logger.warning(f"Could not import Flash Attention enabled models: {e}")
-    FLASH_ATTENTION = False
-
-if FLASH_ATTENTION:
-    __all__.append(FlashNeoXSharded)
-    __all__.append(FlashRWSharded)
-    __all__.append(FlashSantacoderSharded)
-    __all__.append(FlashLlama)
-    __all__.append(FlashGemma)
-    __all__.append(FlashGPT2)
-    __all__.append(FlashQwen)
-    __all__.append(FlashPhi)
-    
-MISTRAL = True
-try:
-    from lorax_server.models.flash_mistral import FlashMistral
-except ImportError as e:
-    logger.warning(f"Could not import Mistral model: {e}")
-    MISTRAL = False
-
-MIXTRAL = True
-try:
-    from lorax_server.models.flash_mixtral import FlashMixtral
-except ImportError as e:
-    logger.warning(f"Could not import Mixtral model: {e}")
-    MIXTRAL = False
-
-if MISTRAL:
-    __all__.append(FlashMistral)
-
-if MIXTRAL:
-    __all__.append(FlashMixtral)
-
 
 def get_model(
     model_id: str,
@@ -145,53 +96,17 @@ def get_model(
             dtypetrust_remote_code=trust_remote_code,
         )
 
-    if model_id.startswith("bigcode/"):
-        if FLASH_ATTENTION:
-            return FlashSantacoderSharded(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        elif sharded:
-            raise NotImplementedError(
-                FLASH_ATT_ERROR_MESSAGE.format("Sharded Santacoder")
-            )
-        else:
-            return SantaCoder(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
+    if model_id.startswith("bigcode/") or model_type == "gpt_bigcode":
+        from lorax_server.models.flash_santacoder import FlashSantacoderSharded
 
-    if model_type == "gpt_bigcode":
-        if FLASH_ATTENTION:
-            return FlashSantacoderSharded(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        elif sharded:
-            raise NotImplementedError(
-                FLASH_ATT_ERROR_MESSAGE.format("Sharded Santacoder")
-            )
-        else:
-            return SantaCoder(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
+        return FlashSantacoderSharded(
+            model_id,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
 
     if model_type == "bloom":
         return BLOOMSharded(
@@ -202,153 +117,105 @@ def get_model(
             dtype=dtype,
             trust_remote_code=trust_remote_code,
         )
+    
     if model_type == "mpt":
         return MPTSharded(
             model_id, revision, quantize=quantize, compile=compile, trust_remote_code=trust_remote_code
         )
 
     if model_type == "gpt_neox":
-        if FLASH_ATTENTION:
-            return FlashNeoXSharded(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        elif sharded:
-            return GPTNeoxSharded(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        else:
-            return CausalLM(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
+        from lorax_server.models.flash_neox import FlashNeoXSharded
+
+        return FlashNeoXSharded(
+            model_id,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
 
     if model_type == "llama":
-        if FLASH_ATTENTION:
-            return FlashLlama(
-                model_id,
-                adapter_id,
-                adapter_source,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        elif sharded:
-            raise NotImplementedError(FLASH_ATT_ERROR_MESSAGE.format("Sharded Llama"))
-        else:
-            return CausalLM(
-                model_id,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
+        from lorax_server.models.flash_llama import FlashLlama
+
+        return FlashLlama(
+            model_id,
+            adapter_id,
+            adapter_source,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
 
     if model_type == "gpt2":
-        if FLASH_ATTENTION:
-            return FlashGPT2(
-                model_id,
-                adapter_id,
-                adapter_source,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        raise NotImplementedError(FLASH_ATT_ERROR_MESSAGE.format("GPT-2"))
+        from lorax_server.models.flash_gpt2 import FlashGPT2
+
+        return FlashGPT2(
+            model_id,
+            adapter_id,
+            adapter_source,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
 
     if model_type in ["RefinedWeb", "RefinedWebModel", "falcon"]:
-        if sharded:
-            if FLASH_ATTENTION:
-                if config_dict.get("alibi", False):
-                    raise NotImplementedError("sharded is not supported for this model")
-                return FlashRWSharded(
-                    model_id,
-                    revision,
-                    quantize=quantize,
-                    compile=compile,
-                    dtype=dtype,
-                    trust_remote_code=trust_remote_code,
-                )
-            raise NotImplementedError(FLASH_ATT_ERROR_MESSAGE.format(f"Sharded Falcon"))
-        else:
-            if FLASH_ATTENTION and not config_dict.get("alibi", False):
-                return FlashRWSharded(
-                    model_id,
-                    revision,
-                    quantize=quantize,
-                    compile=compile,
-                    dtype=dtype,
-                    trust_remote_code=trust_remote_code,
-                )
-            else:
-                return RW(
-                    model_id,
-                    revision,
-                    quantize=quantize,
-                    compile=compile,
-                    dtype=dtype,
-                    trust_remote_code=trust_remote_code,
-                )
+        from lorax_server.models.flash_rw import FlashRWSharded
+
+        return FlashRWSharded(
+            model_id,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
 
     if model_type == "mistral":
-        if MISTRAL:
-            return FlashMistral(
-                model_id,
-                adapter_id,
-                adapter_source,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        raise NotImplementedError("Mistral model requires flash attention v2")
+        from lorax_server.models.flash_mistral import FlashMistral
+
+        return FlashMistral(
+            model_id,
+            adapter_id,
+            adapter_source,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
     
     if model_type == "mixtral":
-        if MIXTRAL:
-            return FlashMixtral(
-                model_id,
-                adapter_id,
-                adapter_source,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        raise NotImplementedError("Mixtral models requires flash attention v2, stk and megablocks")
+        from lorax_server.models.flash_mixtral import FlashMixtral
+
+        return FlashMixtral(
+            model_id,
+            adapter_id,
+            adapter_source,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
     
     if model_type == "qwen":
-        if FLASH_ATTENTION:
-            return FlashQwen(
-                model_id,
-                adapter_id,
-                adapter_source,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        raise NotImplementedError("Qwen model requires flash attention v2")
+        from lorax_server.models.flash_qwen import FlashQwen
+
+        return FlashQwen(
+            model_id,
+            adapter_id,
+            adapter_source,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
 
     if model_type == "qwen2":
         from lorax_server.models.flash_qwen2 import FlashQwen2
@@ -365,32 +232,32 @@ def get_model(
         )
     
     if model_type in ["phi-msft", "phi"]:
-        if FLASH_ATTENTION:
-            return FlashPhi(
-                model_id,
-                adapter_id,
-                adapter_source,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        raise NotImplementedError("Phi model requires flash attention v2")
+        from lorax_server.models.flash_phi import FlashPhi
+
+        return FlashPhi(
+            model_id,
+            adapter_id,
+            adapter_source,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
     
     if model_type == "gemma":
-        if FLASH_ATTENTION:
-            return FlashGemma(
-                model_id,
-                adapter_id,
-                adapter_source,
-                revision,
-                quantize=quantize,
-                compile=compile,
-                dtype=dtype,
-                trust_remote_code=trust_remote_code,
-            )
-        raise NotImplementedError("Gemma model requires flash attention v2")
+        from lorax_server.models.flash_gemma import FlashGemma
+
+        return FlashGemma(
+            model_id,
+            adapter_id,
+            adapter_source,
+            revision,
+            quantize=quantize,
+            compile=compile,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
 
     if model_type == "opt":
         return OPTSharded(
@@ -410,21 +277,6 @@ def get_model(
             compile=compile,
             dtype=dtype,
             trust_remote_code=trust_remote_code,
-        )
-
-    if sharded:
-        raise ValueError("sharded is not supported for AutoModel")
-    if quantize == "gptq":
-        raise ValueError(
-            "gptq quantization is not supported for AutoModel, you can try to quantize it with `lorax-server quantize ORIGINAL_MODEL_ID NEW_MODEL_ID`"
-        )
-    elif (quantize == "bitsandbytes-fp4") or (quantize == "bitsandbytes-nf4"):
-        raise ValueError(
-            "4bit quantization is not supported for AutoModel"
-        )
-    if quantize == "awq":
-        raise ValueError(
-            "awq quantization is not supported for AutoModel"
         )
 
     raise ValueError(f"Unsupported model type {model_type}")

